Kp/www 210 correct dom and attributes

ktp242 opened the pull request:
This PR deals with the display and styles of list selector. It now has proper styles among small and desktop viewports. The title and pick count will display on different components based on different viewports.

The <Select> also has the proper styles and icons as the design shows.

Third, it modifies the rendering of <Hero> and updates the tests.

The design samples can be found here
https://sketch.cloud/s/J88L0/all/staff-picks-v2/mobile-portrait-init-v1
https://sketch.cloud/s/J88L0/all/staff-picks-v2/mobile-portrait-init-v1-dom-overlay
https://sketch.cloud/s/J88L0/all/staff-picks-v2/desktop-hd

EdwinGuzman commented on src/app/components/Sidebar/Sidebar.jsx:
```diff
@@ -54,17 +54,26 @@ const Sidebar = (props) => {
-      {renderListSelector(updateCurrentListSelectorValues(props))}
-      {renderBookFilters(props.isJsEnabled)}
+      <div className="book-filters-heading">
+        <h2 id="list-title" tabIndex="0">
```
You can find this same component in `BookList` at https://github.com/NYPL/staff-picks/blob/development/src/app/components/BookList/BookList.jsx#L25.
I didn't refactor it into its own component but since you're working on it, you should do it.

Also, and very important, you cannot have the same `id` twice on the page and this renders it twice. It might be tricky to focus on one OR the other when switching between mobile/tablet/desktop.

ktp242 replied:
Sure. I thought you were working on it. Didn't want to do the same thing as we did on my last PR.

EdwinGuzman commented on src/app/utils/utils.js:
```diff
@@ -100,6 +100,17 @@ function Utils() {
+   * Makes the first character uppercase in a string
+   * @param {string} str
+   */
+  this.upperCaseFirstChar = (str) => {
```
`capitalize` is a better name.

EdwinGuzman commented on src/client/styles/components/_sidebar.scss:
```diff
@@ -1,18 +1,82 @@
+      -webkit-appearance: none;
+      appearance: none;
+      /* The costmized dropdown icon*/
+      background: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI3LjY5OTA1IiBoZWlnaHQ9IjMuOTg3MTMiIHZpZXdCb3g9IjAgMCA3LjY5OTA1IDMuOTg3MTMiPgogIDx0aXRsZT5kb3duLjIud2VkZ2U8L3RpdGxlPgogIDxwb2x5bGluZSBwb2ludHM9IjcuMzk1IDAuMzYgMy44NSAzLjM3IDAuMzA0IDAuMzU4IiBmaWxsPSIjZmZmIiBzdHJva2U9IiMyMzFmMjAiIHN0cm9rZS1taXRlcmxpbWl0PSIxMCIgc3Ryb2tlLXdpZHRoPSIwLjk0MDUiLz4 ...
```
Where did this come from?

ktp242 replied:
https://nypl.github.io/design-toolkit/sections/forms.html#omniSearch

I copied the same style of the dropdwon from omni search form

EdwinGuzman replied:
I think I see what's happening. You should use the class names and not copy/paste from the toolkit. That should work.

ktp242 replied:
I have some concern about that. Let's talk about this a bit.

ktp242 replied:
yeah I have looked into it. I have some concerns about it.

1. The class names for it are `nypl-omnisearch` and `nypl-omni-fields`. They don't really describe the form correctly for its role.
2. The styles not only include the drop down icon, but also other features of the omni search. So if we use it, we still need to add some other styles to adjust them.

I feel the better way is to customize the styles now and maybe @ricardoom can come up a new style sample for this new dropdown on the toolkit? Then I can switch it.

EdwinGuzman commented:
Also, I'm seeing a few weird things:

1) ~330px and under:
![screen shot 2018-04-30 at 5 37 11 pm](https://user-images.githubusercontent.com/1280564/39451795-f6270e8c-4c9d-11e8-9aab-d363d0c4ca23.png)

2) ~1024px:
![screen shot 2018-04-30 at 5 36 44 pm](https://user-images.githubusercontent.com/1280564/39451809-057a010a-4c9e-11e8-93ed-bbb26fe8b36d.png)

ktp242 commented:
I saw the first case, it happened on my end under 281px. I think we have a smallest viewport we support until? I can try fix it though.

I didn't see the same thing on my end of the second case, but I can go investigate more.

EdwinGuzman commented on src/app/utils/utils.js:
```diff
@@ -102,11 +102,11 @@ function Utils() {
    * @param {string} str
    */
-  this.upperCaseFirstChar = (str) => {
+  this.capitalizeFirstChar = (str) => {
```
Ah, capitalize means making the first char uppercase, so there's no need to also put `FirstChar`.

ktp242 replied:
ah hahaha OK!

ricardoom commented on src/client/styles/components/_sidebar.scss:
```diff
@@ -1,18 +1,82 @@
+    }
+    select {
+      /* Hides -webkit-appearance so we can use costmized dropdown icon*/
+      -webkit-appearance: none;
```
add:
`-moz-appearance: none`

ricardoom replied:
add: 
`border: 1px solid`

ktp242 replied:
Updated!

ktp242 commented:
Updated the branch based on the comments! Also fixed the position of filter's toggle button.

I didn't update the class names for <select> though. As the reason I mentioned above.

EdwinGuzman commented on src/app/components/BookList/BookList.jsx:
```diff
@@ -17,21 +18,18 @@ const BookList = ({
-        <span>{booksfound}</span>
-      </h2>
+      <div className="list-title-container">
+        <ListTitle displayInfo={displayInfo} picksCount={picksCount} idPrefix="" />
```
You don't need `idPrefix=""` if the default value of that prop is an empty string.

ktp242 replied:
updated

EdwinGuzman commented on test/unit/Application.test.js:
```diff
@@ -15,6 +18,82 @@ const params = {
+    let component;
+
+    before(() => {
+      // Not the best way to initialize this component:
```
Why not?

ktp242 replied:
To be honest, I copied the same tests from other `describe` in `Application.test.js` as they will test the same DOMs. I left it as it was the previous dev's comment. I have the same question too.

Please see line 22
https://github.com/NYPL/staff-picks/blob/development/test/unit/Application.test.js

EdwinGuzman commented on test/unit/BookList.test.js:
```diff
@@ -34,9 +35,8 @@ describe('BookList', () => {
-      expect(component.find('h2').length).to.equal(1);
-      expect(component.find('h2').text()).to.equal(' Picks for 0 Books Found');
+    it('should render an <ListTitle> component', () => {
+      expect(component.find(ListTitle).length).to.equal(1);
```
Here you imported the component and then used it inside `.find()`. But in the previous tests, you don't import the components and use them as strings in `.find()`. I prefer importing them and checking for those components. Just be consistent.

ktp242 replied:
Can you point me where the previous tests you see? Do we need to update them in this PR?

ktp242 replied:
Yeah as I mentioned I just reused the same tests from previous dev's tests but expand them to different params. I can update them

EdwinGuzman commented on test/unit/ListTitle.test.js:
```diff
@@ -0,0 +1,70 @@
+    // it('should have an h2', () => {
```
What's up with this??

ktp242 replied:
Sorry this is supposed to be deleted.

ktp242 commented on src/app/utils/utils.js:
```diff
@@ -100,6 +100,17 @@ function Utils() {
+   * capitalizes a string
+   * @param {string} str
+   */
+  this.capitalizeAString = (str) => {
```
Updated

ktp242 commented:
Updated based on the comments!

EdwinGuzman commented on src/app/components/BookList/BookList.jsx:
```diff
@@ -3,35 +3,34 @@ import PropTypes from 'prop-types';
-        <span>{booksfound}</span>
-      </h2>
+      <div className="list-title-container">
+        <ListTitle displayInfo={displayInfo} picksCount={picksCount} idPrefix={idPrefix} />
```
This is not what I meant. Now you added `idPrefix` as a prop dependency for `<BookList>`. You already had the default prop value of `idPrefix` in  `<ListTitle>` as "". So you don't need `idPrefix=""` or `idPrefix={idPrefix}` initialized here at all.

ktp242 replied:
Yup just realized that and updated.

holingpoon commented on appConfig.js:
```diff
@@ -80,25 +80,23 @@ const config = {
-        description: 'Explore our annual selection of outstanding children\'s titles.',
-        heroImageUrl: 'src/client/images/desktop.kids.cover.2017.png',
-      },
+    ya: {
```
If `staffPicks` is getting single quotes, is there any reason why `ya` and `childrens` are not getting single quotes?

ktp242 replied:
ohh it should not be there! Thanks

ktp242 replied:
And updated!
